added `map_word`

ThomasBreuer opened the pull request:
(suggested by @thofma)

The main idea is that one can map an element of a free group to a product of the entries of a given array or of their inverses.

The same is defined for elements of f. p. groups, by delegating to a representative in the underlying free group. This is because in Oscar, these elements have the same type as elements in free groups (and because the same definition is used also in GAP).

The same is defined for a description of elements in free groups by vectors of generator numbers and their negatives ("letter representation") or by vectors of pairs `i => e` denoting the `e`-th power of the `i`-th generator ("syllable representation").

Note:
- GAP allows one to specify in the construction of a free group whether its elements will be represented via letters or syllables. Up to now, there is no such choice in Oscar.
- The GAP function `MappedWord` is optimized for the (common?) situation that the given images are themselves elements of the free group. This case is ignored in the current pull request.
- The syntax of `MappedWord` is affected by this perspective in the sense that in GAP, the identity of the free group on 0 generators is mapped to itself if one prescribes an empty list of images for the empty list of generators. Here I propose a different solution, using an optional argument `init` for this situation.
- In order to avoid computing the inverses of generators over and over again, `map_word` supports an array containing (some) inverses of the given images, and if new inverses are needed then they will be stored in this array, in order to be available in the next calls of `map_word` for the same images. This seems to be more flexible than the GAP syntax of `MappedWord`, where one has to enter also the list of generators, and where it is allowed to extend this list as well as the list of images by inverses.

ThomasBreuer committed:
added `map_word`

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1332,6 +1333,128 @@ function relators(G::FPGroup)
+    # element of a f.p. group
+    gX = GAP.Globals.UnderlyingElement(gX)
+  end
+  _names = GAP.evalstr("x -> x!.names")
```
With this, each time `map_word` is called, we again run this code (so the GAP kernel parses that string, turns it into an abstract syntax tree and returns yet another fresh function. If one calls `map_word` a lot, this will quickly add up.

At the same time, this is "only" used in an assertion, which makes this doubly painful.

To resolve this, we can put this into a global variable, like so:

```julia
# on the global level
const _namesGAP = Ref{GapObj}()

...

   # in the function, or in code called by __init__
   if !isassigned(_namesGAP)
     _namesGAP[] = GAP.evalstr("x -> x!.names")
   end

```

ThomasBreuer replied:
I had opened the pull request oscar-system/GAP.jl/pull/816 in order to get a better solution. Meanwhile it got merged in GAP.jl, and the code lines in question can be replaced when the functionality becomes available in Oscar.

fingolfin replied:
OK, GAP.jl 0.8.1 is out, so you could change this PR to require it in the `Project.toml` and use your new accessor from here

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1332,6 +1333,128 @@ function relators(G::FPGroup)
+    ll = Vector{Int}(GAP.Globals.LetterRepAssocWord(gX))
+  elseif GAP.Globals.IsSyllableAssocWordRep(gX)
+    l = ExtRepOfObj(gX)
+    ll = Pairs{Int, Int}[]
```
Syntax error:
```suggestion
    ll = Pair{Int, Int}[]
```
This code branch is not covered by a test, which is why this could slip through :-)

ThomasBreuer replied:
Yes. I am going to add suitable tests.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1332,6 +1333,128 @@ function relators(G::FPGroup)
+#TODO: improve this, see https://github.com/oscar-system/GAP.jl/issues/771
+  @assert GAP.Globals.IsAssocWord(gX)
+  if GAP.Globals.IsLetterAssocWordRep(gX)
+    ll = Vector{Int}(GAP.Globals.LetterRepAssocWord(gX))
```
Why not always use `ExtRepOfObj` ? It is defined to for all word reps?

ThomasBreuer replied:
Hmm.
`ExtRepOfObj` returns a syllable description. For words that are described internally by letters, calling `ExtRepOfObj` means to compute this description in GAP. This is unnecessary because Julia anyhow wants to support also the letter representation.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1332,6 +1333,128 @@ function relators(G::FPGroup)
+    ll = Pairs{Int, Int}[]
+    for i in 1:2:length(l)
+      push!(ll, l[i] => l[i+1])
+    end
```
Might also do
```suggestion
    ll = [l[i] => l[i+1] for i in 1:2:length(l)]
```

ThomasBreuer replied:
Yes, thanks.

thofma commented:
Thanks, I like it.

ThomasBreuer committed:
incorporate the hints for improvements

fingolfin commented on test/Groups/homomorphisms.jl:
```diff
@@ -96,6 +96,82 @@ end
+           GAP.Globals.ElementsFamily(GAP.Globals.FamilyObj(FL.X)))
+   for F in [FL, FS]
+     F1 = gen(F, 1)
+     F2 = gen(F, 2)
```
Fun fact: you can also do this:
```suggestion
     F1, F2 = gens(F)
```

ThomasBreuer committed:
adjust to latest GAP.jl

- use `GAP.getbangproperty` in `map_word
- for that, require version 0.8.1 of GAP.jl
- improve the syntax of some tests

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1332,6 +1333,126 @@ function relators(G::FPGroup)
+  elseif GAP.Globals.IsSyllableAssocWordRep(gX)
+    # Here we take the available syllable representation.
+    l = GAP.Globals.ExtRepOfObj(gX)::GAP.GapObj
+    ll = Pair{Int, Int}[l[i] => l[i+1] for i in 1:2:length(l)]
```
I should note that in principle, exponents that don't fit into an Int could occur. But I don't think GAP deals very well with those right now, and in any case, it'll lead to an exception if they appear, and we can deal with it when someone reports that

ThomasBreuer replied:
Concerning the statement that GAP does not deal very well with words with huge exponents:
I have not been able to produce such a group element by powering up a suitable generator.
The `SIZE_OBJ` value of the result would be small, but intermediately the process runs out of space.

thofma commented:
Might also be interesting for @ulthiel.
